pennylane/templates/subroutines/amplitude_amplification.py

```diff
@@ -0,0 +1,184 @@
+
+    .. math::
+
+            \text{A}(U, O)|\Psi\rangle \sim |\phi\rangle.
+
+    The implementation of the algorithm is based on `[arXiv:quant-ph/0005055] <https://arxiv.org/abs/quant-ph/0005055>`__.
+    The template also unlocks advanced techniques such as fixed-point quantum search [`arXiv:1409.3305 <https://arxiv.org/abs/1409.3305>`__] and oblivious amplitude amplification [`arXiv:1312.1414 <https://arxiv.org/abs/1312.1414>`__] by reflecting on a subset of the wires.
+
+    Args:
+        U (Operator): operator that prepares the state :math:`|\Psi\rangle`
+        O (Operator): the oracle that flips the sign of the state :math:`|\phi\rangle` and does nothing to the state :math:`|\phi^{\perp}\rangle`
+        iters (int): the number of iterations of the amplitude amplification subroutine, default is ``1``
+        fixed_point (bool): whether to use the fixed-point amplitude amplification algorithm, default is ``False``
+        work_wire (int): the auxiliary wire to use for the fixed-point amplitude amplification algorithm, default is ``None``
+        reflection_wires (Wires): the wires to reflect on, default is the wires of ``U``
+        p_min (int): the lower bound for the probability of success in fixed-point amplitude amplification, default is ``0.9``
+
+    Raises:
+        ValueError: ``work_wire`` must be specified if ``fixed_point == True``.
+        ValueError: ``work_wire`` must be different from the wires of the oracle ``O``.
+
+    **Example**
+
+    Amplification of state :math:`|2\rangle` using Grover's algorithm with 3 qubits.
+    The state :math:`|\Psi\rangle` is constructed as a uniform superposition of basis states.
+
+    .. code-block::
+
+        @qml.prod
+        def generator(wires):
+          for wire in wires:
+            qml.Hadamard(wires = wire)
+
+        U = generator(wires = range(3))
+        O = qml.FlipSign(2, wires = range(3))
+
+        dev = qml.device("default.qubit")
+
+        @qml.qnode(dev)
+        def circuit():
+
+          generator(wires = range(3))
+          qml.AmplitudeAmplification(U, O, iters = 5, fixed_point=True, work_wire=3)
+
+          return qml.probs(wires = range(3))
+
+    .. code-block:: pycon
+
+        >>> print(np.round(circuit(),3))
+        [0.009 0.009 0.94  0.009 0.009 0.009 0.009 0.009]
+    """
+
+    def _flatten(self):
+        data = (self.hyperparameters["U"], self.hyperparameters["O"])
+        metadata = tuple(
+            value for key, value in self.hyperparameters.items() if key not in ["O", "U"]
+        )
+        return data, metadata
+
+    @classmethod
+    def _unflatten(cls, data, metadata):
+        U, O = (data[0], data[1])
+        return cls(
+            U,
+            O,
+            iters=metadata[0],
+            fixed_point=metadata[1],
+            work_wire=metadata[2],
+            p_min=metadata[3],
+            reflection_wires=metadata[4],
+        )
```

Comment:
This makes me a little nervous because if the order in the dictionary changes for some reason, we will be mixing up the metadata entires with each other. My personal suggestion would be to use this: 

```suggestion
    def _flatten(self):
        data = (self.hyperparameters["U"], self.hyperparameters["O"])
        metadata = tuple(
            (key, value) for key, value in self.hyperparameters.items() if key not in ["O", "U"]
        )
        return data, metadata

    @classmethod
    def _unflatten(cls, data, metadata):
        U, O = (data[0], data[1])
        hyperparams_dict = dict(metadata)
        return cls(U, O, **hyperparams_dict)
```